doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,352 @@
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+With NVM
+--------
+With NVM, we are able to make use of a different scheme to store the data depending replication
+performance.
+All of the data store NVM first with out-of-place update manner, then flush them to NVMe storage device
+using in-place update scheme. This is feasible when replication latency (> 0.5 ms) is slower than
+NVMe storage performance (< 0.5 ms) because flush can store all buffered data in NVM in batch-manner.
+
+
+Detailed Design 
+===============
+
+
+- Onode lookup
+Upon onode lookup requests
+First, we hash CID (collection ID) to get the hash index of the hash table. Note that the entire hash table can be 
+pre-allocated when we limit the maximum number of objects that can be created in an AG. With the hash index, 
+the buckets are scanned until it finds the hash entry for the target object ID.
+A hash entry contains <CID, OID, Type, Offset, Segment>.
+With the hash entry, we can find the disk location of the target onode.
+Since the OID meta entry in the segment contains <OID, Chunk Number, (Offset, len), (Offset, len), … >, 
+we can read the onode from the target chunk.
+
+- Allocation
+Allocation Groups
+Entire disk space is divided into a number of equally sized chunks called Allocation Groups (AG).
+Each reactor thread in crimson-osd is responsible for a set of AGs (# of AGs / # of threads).
+Each AG has its own data structures to manage the disk partition.
+
+- Data allocation
+As we explained above, the entire hash table is pre-allocated in each AG. Given CID and OID, we can map any data to 
+the segment and offset. A segment is comprised of contiguous blocks. Blocks can be allocated by searching the free space 
+tracking data structure (we explain below).
+Based on onode lookup, we probably know where OID is located via segment number. And, segment contains multiple entries 
+such as oid meta entries (object_info_t, data etc.), small data, and data. (segment header describe what entries exists there).
+So a single write can store a transaction data at once.
```

Comment:
I don't quite follow.  What do you mean by "probably know where OID is located via segment number"?  If you always pack all information for a write into a single segment, how is it possible for the data for two arbitrary oids to be mutated in the same transaction?  Are you simply describing the journal entry here with actual data writeout to happen lazily?  Are all transactions (on the whole PS?  on the reactor?) written to the same segment until it ends?

Reply:
> Are you simply describing the journal entry here with actual data writeout to happen lazily?

Yes.